13_BitonicSort/main.cpp

```diff
@@ -0,0 +1,323 @@
+			bufferData[0][i] = g() % 10000; 
+		}
+
+		memcpy(mapped_memory[0], bufferData[0], sizeof(uint32_t) * element_count);
+
+		for (uint32_t i = 0; i < element_count; i++) {
+			bufferData[1][i] = i; // Values are indices for verification
+		}
+
+		memcpy(mapped_memory[1], bufferData[1], sizeof(uint32_t) * element_count);
+
+		std::string outBuffer;
+
+		outBuffer.append("ALL ELEMENTS: ");
+		for (auto i = 0; i < element_count; i++) {
+			outBuffer.append("{");
+			outBuffer.append(std::to_string(bufferData[0][i]));
+			outBuffer.append(", ");
+			outBuffer.append(std::to_string(bufferData[1][i]));
+			outBuffer.append("} ");
+
+			// Add newline every 20 elements for readability
+			if ((i + 1) % 20 == 0) {
+				outBuffer.append("\n");
+			}
+		}
+		outBuffer.append("\n");
+		outBuffer.append("Count: ");
+		outBuffer.append(std::to_string(element_count));
+		outBuffer.append("\n");
+		m_logger->log("Your input array is: \n" + outBuffer, ILogger::ELL_PERFORMANCE);
+
+
+		smart_refctd_ptr<nbl::video::IGPUCommandBuffer> cmdBuf;
+		{
+			smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(getComputeQueue()->getFamilyIndex(), IGPUCommandPool::CREATE_FLAGS::RESET_COMMAND_BUFFER_BIT);
+			if (!cmdpool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, 1u, &cmdBuf))
+				return logFail("Failed to create Command Buffers!\n");
+		}
+
+		constexpr uint64_t started_value = 0;
+		uint64_t timeline = started_value;
+		smart_refctd_ptr<ISemaphore> progress = m_device->createSemaphore(started_value);
+
+		auto pc = BitonicPushData{
+			.inputKeyAddress = buffer_device_address[0],
+			.inputValueAddress = buffer_device_address[1],
+			.outputKeyAddress = buffer_device_address[2],
+			.outputValueAddress = buffer_device_address[3],
+			.dataElementCount = element_count
+		};
+
+		cmdBuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+		cmdBuf->beginDebugMarker("Bitonic Sort Single Dispatch", core::vectorSIMDf(0, 1, 0, 1));
+		cmdBuf->bindComputePipeline(bitonicSortPipeline.get());
+		cmdBuf->pushConstants(layout.get(), IShader::E_SHADER_STAGE::ESS_COMPUTE, 0u, sizeof(pc), &pc);
+		cmdBuf->dispatch(1, 1, 1);
+		cmdBuf->endDebugMarker();
+		cmdBuf->end();
+
+		{
+			auto queue = getComputeQueue();
+
+			IQueue::SSubmitInfo submit_infos[1];
+			IQueue::SSubmitInfo::SCommandBufferInfo cmdBufs[] = {
+				{
+					.cmdbuf = cmdBuf.get()
+				}
+			};
+			submit_infos[0].commandBuffers = cmdBufs;
+			IQueue::SSubmitInfo::SSemaphoreInfo signals[] = {
+				{
+					.semaphore = progress.get(),
+					.value = ++timeline,
+					.stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT
+				}
+			};
+			submit_infos[0].signalSemaphores = signals;
+
+			m_api->startCapture();
+			queue->submit(submit_infos);
+			m_api->endCapture();
+		}
+
+		const ISemaphore::SWaitInfo wait_infos[] = { {
+				.semaphore = progress.get(),
+				.value = timeline
+			} };
+		m_device->blockForSemaphores(wait_infos);
```

Comment:
add a separate benchmark mode that does more workgroup dispatches (but allocates the buffer on DeviceLocal without redability) and on a loop with a time elapsed query

If you're curious you can see if you're faster than https://github.com/AndrewBoessen/Bitonic-Merge-Sort